third_party/move/move-vm/runtime/src/native_functions.rs

```diff
@@ -209,19 +205,49 @@ impl<'b> NativeContext<'_, 'b> {
         self.interpreter.get_stack_frames(count)
     }
 
-    pub fn gas_balance(&self) -> InternalGas {
-        self.gas_balance
+    pub fn legacy_gas_budget(&self) -> InternalGas {
+        self.gas_meter.legacy_gas_budget_in_native_context()
+    }
+
+    pub fn use_heap_memory(&mut self, amount: u64) -> PartialVMResult<()> {
+        if self.is_metering_in_native_context_enabled() {
+            self.gas_meter.charge_heap_memory(amount)?;
+        } else {
+            self.legacy_heap_memory_usage = self.legacy_heap_memory_usage.saturating_add(amount);
+        }
+        Ok(())
+    }
+
+    pub fn legacy_heap_memory_usage(&self) -> u64 {
+        self.legacy_heap_memory_usage
+    }
+
+    /// Returns the gas meter used for execution. Even if native functions cannot use it to
+    /// charge gas (feature-gating), gas meter can be used to query gas meter's balance.
+    pub fn gas_meter(&mut self) -> &mut dyn NativeGasMeter {
+        self.gas_meter
     }
 
-    pub fn use_heap_memory(&mut self, amount: u64) {
-        self.heap_memory_usage = self.heap_memory_usage.saturating_add(amount);
+    /// Returns true if native functions have access to gas meter and cah charge gas. Otherwise,
+    /// only VM's interpreter needs to charge gas.
+    pub fn is_metering_in_native_context_enabled(&self) -> bool {
+        self.gas_meter.is_metering_in_native_context_enabled()
     }
```

Comment:
Similarly, I feel like this is a leaky abstraction -- the Move VM `NativeContext` does not need to know if we have direct gas meter access as it now always passes through the gas meter to the `SafeNativeContext`

Reply:
Removed, thanks!